src/core/util/http.cc

```diff
@@ -44,63 +42,70 @@
+#include "client/client_context.h"
+#include "client/address_book.h"
 #include "util/filesystem.h"
 
 namespace i2p {
 namespace util {
-namespace http {
+namespace http {  // TODO(anonimal): consider removing this namespace (its not needed)
 
-bool URI::Parse(
+// TODO(unassigned): currently unused but will be useful
+// without needing to create a new object for each given URI
+bool HTTP::Download(
     const std::string& uri) {
-  boost::network::uri::uri URI(uri);
-  if (URI.is_valid()) {
-    m_Protocol.assign(URI.scheme());
-    m_Host.assign(URI.host());
-    m_Port.assign(URI.port());
-    m_Path.assign(URI.path());
-    m_Query.assign(URI.query());
-    // Set defaults for AddressBook
-    // TODO(anonimal): this should disappear once we finish other refactoring
-    if (!m_Port.empty())
-      return true;
-    if (m_Protocol == "https") {
-      m_Port.assign("443");
-    } else {
-      m_Port.assign("80");
-    }
-    return true;
-  } else {
+  SetURI(uri);
+  return Download();
+}
+
+bool HTTP::Download() {
+  if (!GetURI().is_valid()) {
     LogPrint(eLogError, "URI: invalid URI");
     return false;
   }
+  // TODO(anonimal): ideally, we simply swapout the request/response handler
+  // with cpp-netlib so we don't need two separate functions
+  if (!HostIsI2P())
+    return DownloadViaClearnet();
+  return DownloadViaI2P();
 }
 
-// TODO(anonimal): research if cpp-netlib can do this better
-// Used by HTTPProxy
-std::string URI::Decode(
-    const std::string& data) {
-  std::string res(data);
-  for (size_t pos = res.find('%');
-      pos != std::string::npos;
-      pos = res.find('%', pos + 1)) {
-    const char c = strtol(res.substr(pos + 1, 2).c_str(), NULL, 16);
-    res.replace(pos, 3, 1, c);
+bool HTTP::HostIsI2P() {
+  auto uri = GetURI();
+  if (uri.host().substr(uri.host().size() - 4) == ".i2p") {
+    if (!uri.port().empty())
+      return true;
+    // We must assign a port if none was assigned (for internal reasons)
+    std::string port;
+    if (uri.scheme() == "https")
+      port.assign("443");
+    else
+      port.assign("80");
+    // If user supplied user:password, we must append @
+    std::string user_info;
+    if (!uri.user_info().empty()) {
+      user_info.assign(uri.user_info() + "@");
+    }
+    // TODO(anonimal): easier way with cpp-netlib?
+    std::string new_uri(
+        uri.scheme() + "://" + user_info
+        + uri.host() + ":" + port
+        + uri.path() + uri.query() + uri.fragment());
+    SetURI(new_uri);
+    return true;
+  } else {
+    return false;
   }
-  return res;
 }
 
-bool HTTP::Download(
-    const std::string& address) {
-  // Validate URI
-  URI uri;
-  if (!uri.Parse(address))
-    return false;
-  namespace http = boost::network::http;
-  namespace network = boost::network;
-  http::client::options options;
+bool HTTP::DownloadViaClearnet() {
+  auto uri = GetURI();
+  // Create and set options
+  Options options;
+  options.timeout(45);  // Java I2P defined
```

Comment:
It could but, unless its reused elsewhere (used more than once) or part of a bigger ideological container, I don't see the point.